### ISSUE TITLE
torch.cuda.amp.GradScaler._scale can overflow

### ISSUE THREAD

**Issue by Parskatt** [pytorch/pytorch#98828]
### 🐛 Describe the bug

For networks where the loss is small, it can happen that the gradscaler overflows before the gradients become infinite. 
```python
import torch
import torch.nn as nn

net = nn.Linear(5,1).cuda()
optimizer = torch.optim.Adam(net.parameters())
scaler = torch.cuda.amp.GradScaler(growth_interval=1)
for t in range(1000):
    optimizer.zero_grad()
    x = torch.randn(1,5).cuda()
    y = 0.0001 * torch.randn(1,1).cuda()
    l = ((net(x)-y)**2).mean()
    scaler.scale(l).backward()
    scaler.step(optimizer)
    scaler.update()
    print(scaler._scale)
```

Limiting the growth of the gradscaler, or using float64 instead of float32 seems like a reasonable idea.

### Versions

Versions of relevant libraries:
[pip3] numpy==1.24.2
[pip3] torch==2.0.0
[pip3] torchvision==0.15.1
[pip3] triton==2.0.0

cc @mcarilli @ptrblck @leslie-fang-intel @jgong5

**Comment by ngimel** [comment]
cc @crcrpar @janeyx99

**Comment by janeyx99** [comment]
@Parskatt I've opened a PR that will cap the scale growth, but my general sense is that if the scale has gone that large, something else may have already gone wrong and this new lack of warning may only prolong one's struggle. Proceed with caution!

**Comment by Parskatt** [comment]
I think that is a better solution, yes :)

**Comment by janeyx99** [comment]
@Parskatt Notifying you about some relevant discussion in https://github.com/pytorch/pytorch/issues/99640 about adding an assert on the scale. There are nontrivial perf hits with adding an assert on the scale value which makes it not necessarily a good solution for all.

In the meantime, would you rather me land the capping of the scale growth PR or leave the current state as is?

**Comment by Parskatt** [comment]
I think capping makes sense :)
I think that behaviour makes more sense than quietly overflowing.

**Comment by janeyx99** [comment]
Okay, I'll work on landing the PR today c:
